libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/UnitTestTools.hs

```diff
@@ -0,0 +1,184 @@
+import Cardano.Ledger.UMap (RDPair (..))
+import Cardano.Ledger.UTxO (UTxO (..))
+import Cardano.Ledger.Val (Val (..))
+import Data.Default.Class (Default (def))
+import Data.Kind (Type)
+import GHC.TypeNats (KnownNat, Nat)
+import Lens.Micro
+import Test.Cardano.Ledger.Core.Arbitrary ()
+import Test.Cardano.Ledger.Core.KeyPair (KeyPair (..), mkWitnessVKey)
+import Test.QuickCheck
+
+-- ==================================================
+
+-- | Supports patterns like
+--   (tom, john, ann, ron, mary) <- fresh @5 @(Credential 'Staking (EraCrypto era))
+class KnownNat n => Fresh (n :: Nat) (a :: Type) where
+  type Tuple n a :: Type
+  fresh :: Arbitrary a => Tuple n a
+
+instance Fresh 1 a where
+  type Tuple 1 a = Gen a
+  fresh = arbitrary @a
+
+instance Fresh 2 a where
+  type Tuple 2 a = Gen (a, a)
+  fresh = (arbitrary @(a, a))
+
+instance Fresh 3 a where
+  type Tuple 3 a = Gen (a, a, a)
+  fresh = (arbitrary @(a, a, a))
+
+instance Fresh 4 a where
+  type Tuple 4 a = Gen (a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a))
+
+instance Fresh 5 a where
+  type Tuple 5 a = Gen (a, a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a, a))
+
+instance Fresh 6 a where
+  type Tuple 6 a = Gen (a, a, a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a, a, a))
+
+instance Fresh 7 a where
+  type Tuple 7 a = Gen (a, a, a, a, a, a, a)
+  fresh = (arbitrary @(a, a, a, a, a, a, a))
+
+genLens :: Gen a -> Lens' a b -> b -> Gen a
+genLens gen l b = do a <- gen; pure (a & l .~ b)
+
+arbitraryLens :: Arbitrary a => Lens' a b -> b -> Gen a
+arbitraryLens = genLens arbitrary
+
+-- | Use this when you know the TxBody Hash
+keyHashWitVKeyPairFor ::
+  forall kr.
+  SafeHash StandardCrypto EraIndependentTxBody ->
+  Gen (KeyHash kr StandardCrypto, WitVKey 'Witness StandardCrypto)
+keyHashWitVKeyPairFor safehash = do
+  keyPair@(KeyPair vk _) <- arbitrary @(KeyPair kr StandardCrypto)
+  let witness = mkWitnessVKey safehash keyPair
+      keyhash = hashKey vk
+  pure (keyhash, witness)
+
+-- | Use this when any TxBody hash will do
+--   You just need a KeyHash and its Witness
+someKeyHashWitVKeyPair ::
+  forall kr. Gen (KeyHash kr StandardCrypto, WitVKey 'Witness StandardCrypto)
+someKeyHashWitVKeyPair = do
+  safehash <- arbitrary @(SafeHash StandardCrypto EraIndependentTxBody)
+  keyHashWitVKeyPairFor safehash
+
+addr :: Credential 'Staking c -> Maybe (Credential 'Staking c) -> Addr c
+addr cr Nothing = Addr Testnet (coerceKeyRole cr) StakeRefNull
+addr c (Just d) = Addr Testnet (coerceKeyRole c) (StakeRefBase d)
+
+txOut :: forall era. EraTxOut era => Addr (EraCrypto era) -> Coin -> TxOut era
+txOut add coinx = mkBasicTxOut add (modifyCoin (const coinx) zero)
+
+rewardAccount :: Credential 'Staking c -> RewardAccount c
+rewardAccount cred = RewardAccount Testnet cred
+
+emptyPState :: PState era
+emptyPState = def
+
+emptyDState :: DState era
+emptyDState = def
+
+emptyVState :: VState era
+emptyVState = def
+
+emptyCertState :: CertState era
+emptyCertState = def
```

Comment:
So there is a reason for this.  Suppose you have
data PState era
  = PState {psStakePoolParams :: !(Map
                                     (KeyHash 'StakePool (EraCrypto era))
                                     (PoolParams (EraCrypto era))),
            psFutureStakePoolParams :: !(Map
                                           (KeyHash 'StakePool (EraCrypto era))
                                           (PoolParams (EraCrypto era))),
            psRetiring :: !(Map
                              (KeyHash 'StakePool (EraCrypto era))
                              Cardano.Slotting.Slot.EpochNo),
            psDeposits :: !(Map (KeyHash 'StakePool (EraCrypto era)) Coin)}
and State has a Default instance.

Now if you write
def{ppId = x}

Haskell  cannot type check this, because era is ambiguous.  This is because the data for PState never mentions era, only . (EraCrypto era)

So one has to put a type on def. But there is no room to put a type on def. because the haskell parser does not let you write
def @(PState era) {def{ppId = x}
So emptyPState avoids this.  
So to avoid this is use  arbrtraryLens instead.

Reply:
> So one has to put a type on def. But there is no room to put a type on def. because the haskell parser does not let you write
def @(PState era) {def{ppId = x}
So emptyPState avoids this.

There is an easy solution to this, just wrap the expression in parenthesis:

```haskell
(def @(PState era)) {ppId = x}
```